### PR TITLE
Add `LiteLLM`

### LINKED ISSUE #434: Add `LiteLLM` as `LLM`
**Is your feature request related to a problem? Please describe.**
BerriAI LiteLLM can function as a generalized bridge API for a lot of other APIs when we support it.

**Describe the solution you'd like**
A basic implementation for this. 

**Describe alternatives you've considered**
N.A.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by davidberenstein1957**
Closes #434 

```python
import time

from distilabel.llm.litellm import LitellmLLM

if __name__ == "__main__":
    for model in ["mistral/mistral-tiny", "gpt-4"]:
        start_time = time.time()
        llm = LitellmLLM(model=model)
        llm.load()
        output = llm.generate(
            inputs=[
                [
                    {"role": "system", "content": ""},
                    {
                        "role": "user",
                        "content": "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    },
                ]
            ]
        )
        print("--- %s seconds ---" % (time.time() - start_time))
```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -40,6 +40,7 @@ llama-cpp = ["llama-cpp-python >= 0.2.0"]
 ollama = ["ollama >= 0.1.4"]
 openai = ["openai >= 1.0.0"]
 vllm = ["vllm >= 0.2.1"]
+litellm = ["litellm >= 1.30.0"]
```
Comment (alvarobartt) [inline pyproject.toml:43]: These should be alphabetically sorted

File: src/distilabel/llm/litellm.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+from pydantic import PrivateAttr
+
+from distilabel.llm.base import AsyncLLM
+
+if TYPE_CHECKING:
+    from distilabel.llm.typing import GenerateOutput
+    from distilabel.steps.task.typing import ChatType
+
+
+class LitellmLLM(AsyncLLM):
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:27]: To me `LitellmLLM` is redundant, so I'd go with something like `LiteLLM` and that's it, similarly to what we did with `vLLM`

Comment (alvarobartt) [reply src/distilabel/llm/litellm.py:27]:
```suggestion
class LiteLLM(AsyncLLM):
```

File: src/distilabel/llm/litellm.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    """LiteLLM LLM implementation running the async API client.
+
+    Attributes:
+        model: the model name to use for the LLM e.g. "gpt-3.5-turbo", "gpt-4", etc.
+        api_key: the API key to authenticate the requests to the OpenAI API.
+    """
+
+    model: str
+    litellm_logging: Optional[bool] = False
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:36]:
```suggestion
    verbose: bool = False
```

File: src/distilabel/llm/litellm.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    from distilabel.llm.typing import GenerateOutput
+    from distilabel.steps.task.typing import ChatType
+
+
+class LitellmLLM(AsyncLLM):
+    """LiteLLM LLM implementation running the async API client.
+
+    Attributes:
+        model: the model name to use for the LLM e.g. "gpt-3.5-turbo", "gpt-4", etc.
+        api_key: the API key to authenticate the requests to the OpenAI API.
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:32]: These needs to be updated

File: src/distilabel/llm/litellm.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    """
+
+    model: str
+    litellm_logging: Optional[bool] = False
+    _aclient: Optional["callable"] = PrivateAttr(...)
+
+    def load(
+        self,
+    ) -> None:
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:41]:
```suggestion
    def load(self, api_key: Optional[str] = None) -> None:
```

File: src/distilabel/llm/litellm.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        """
+        Loads the `AsyncOpenAI` client to benefit from async requests.
+
+        Args:
+            model: the model name to use for the LLM e.g. "gpt-3.5-turbo", "gpt-4", etc.
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:46]: This docstring also needs an update

```suggestion
        Loads the `AsyncOpenAI` client to benefit from async requests.
```

File: src/distilabel/llm/litellm.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            import litellm
+            from litellm import acompletion
+
+            self._aclient = acompletion
+            litellm.telemetry = False
+        except ImportError as e:
+            raise ImportError(
+                "LiteLLM Python client is not installed. Please install it using"
+                " `pip install litellm`."
+            ) from e
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:59]:
```suggestion
        try:
            import litellm

            litellm.telemetry = False
        except ImportError as e:
            raise ImportError(
                "LiteLLM Python client is not installed. Please install it using"
                " `pip install litellm`."
            ) from e

        self._aclient = litellm.acompletion
```

File: src/distilabel/llm/litellm.py (hunk 5 above)
Comment (alvarobartt) [reply src/distilabel/llm/litellm.py:41]: Maybe the `pre-commit` is not installed / updated?

File: src/distilabel/llm/litellm.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+        if not self.litellm_logging:
+            for key in logging.Logger.manager.loggerDict.keys():
+                if "litellm" in key.lower():
+                    logger = logging.getLogger(key)
+                    logger.setLevel(logging.ERROR)
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:65]:
```suggestion
        if not self.verbose:
            for key in logging.Logger.manager.loggerDict.keys():
                if "litellm" not in key.lower():
                    continue
                logging.getLogger(key).setLevel(logging.CRITICAL)
```

File: src/distilabel/llm/litellm.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+                Defaults to `600`.
+            custom_llm_provider: Used for Non-OpenAI LLMs, Example usage for bedrock, set(iterable)
+                model="amazon.titan-tg1-large" and custom_llm_provider="bedrock". Defaults to
+                `None`.
+
+        Returns:
+            A list of lists of strings containing the generated responses for each input.
+        """
+
+        completion = await self._aclient(  # type: ignore
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:137]: Does it support `n`? Some do not as e.g. `Mistral`, so could you check that the generations are consistent when running some models other than the OpenAI ones? Thanks in advance!

We've seem some issues with the `n` paramater, and wanted to double check how is `LiteLLM` handling that

File: tests/unit/llm/test_litellm.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            choices=[Mock(message=Mock(content=" Aenean hendrerit aliquam velit. ..."))]
+        )
+        llm._aclient.chat.completions.create = AsyncMock(return_value=mocked_completion)
+
+        nest_asyncio.apply()
+
+        llm.generate(
+            inputs=[
+                [
+                    {"role": "system", "content": ""},
```
Comment (alvarobartt) [inline tests/unit/llm/test_litellm.py:65]: Does this mean that the `system` message is required by `LiteLLM`?

```suggestion
                    {"role": "system", "content": ""},
```

**Review by alvarobartt — CHANGES_REQUESTED** [review]

File: src/distilabel/llm/litellm.py (hunk 5 above)
Comment (davidberenstein1957) [reply src/distilabel/llm/litellm.py:41]: it is, perhaps it doesn't register due to the comma?

File: src/distilabel/llm/litellm.py (hunk 9 above)
Comment (davidberenstein1957) [reply src/distilabel/llm/litellm.py:137]: Sure

File: src/distilabel/llm/litellm.py (hunk 4 above)
Comment (alvarobartt) [reply src/distilabel/llm/litellm.py:32]: I've check and this still needs an update, so I'll unresolve the conversation

File: src/distilabel/llm/litellm.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    """
+
+    model: str
+    verbose: bool = False
+    _aclient: Optional["callable"] = PrivateAttr(...)
+
+    def load(self) -> None:
+        """
+        Loads the `AsyncOpenAI` client to benefit from async requests.
+        """
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:43]: To be updated

File: src/distilabel/llm/litellm.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+            import litellm
+            from litellm import acompletion
+
+            litellm.telemetry = False
+        except ImportError as e:
+            raise ImportError(
+                "LiteLLM Python client is not installed. Please install it using"
+                " `pip install litellm`."
+            ) from e
+        self._aclient = acompletion
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:55]:
```suggestion
            import litellm

            litellm.telemetry = False
        except ImportError as e:
            raise ImportError(
                "LiteLLM Python client is not installed. Please install it using"
                " `pip install litellm`."
            ) from e
        self._aclient = litellm.acompletion
```

File: src/distilabel/llm/litellm.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                    mock_response=mock_response,
+                    force_timeout=force_timeout,
+                    custom_llm_provider=custom_llm_provider,
+                )
+                choices.extend(completion.choices)
+            return choices
+
+        try:
+            litellm.drop_params = False
+            choices = await _call_aclient_untill_n_choices()
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:167]: Typo `untill`->`until`

```suggestion
            choices = await _call_aclient_until_n_choices()
```

File: src/distilabel/llm/litellm.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        except litellm.exceptions.APIError as e:
+            if "does not support parameters" in str(e):
+                litellm.drop_params = True
+                choices = await _call_aclient_untill_n_choices()
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:171]: Typo `untill`->`until`

```suggestion
                choices = await _call_aclient_until_n_choices()
```

File: src/distilabel/llm/litellm.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+            custom_llm_provider: Used for Non-OpenAI LLMs, Example usage for bedrock, set(iterable)
+                model="amazon.titan-tg1-large" and custom_llm_provider="bedrock". Defaults to
+                `None`.
+
+        Returns:
+            A list of lists of strings containing the generated responses for each input.
+        """
+        import litellm
+
+        async def _call_aclient_untill_n_choices():
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:135]: Missing return type-hint, assuming something like `-> List["Choices"]`

File: tests/unit/llm/test_litellm.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+import pytest
+from distilabel.llm.litellm import LiteLLM
+
+
+@patch("litellm.acompletion")
+class TestLiteLLM:
+    def test_litellm_llm(self, _: MagicMock) -> None:
+        llm = LiteLLM(model="gpt-4", api_key="api.key")  # type: ignore
+        assert isinstance(llm, LiteLLM)
+        assert llm.model_name == "gpt-4"
```
Comment (alvarobartt) [inline tests/unit/llm/test_litellm.py:28]: Could you add more tests with the different providers that `LiteLLM` offers? i.e. one function per provider? or at least the most common ones

```suggestion
    def test_litellm_llm(self, _: MagicMock) -> None:
        llm = LiteLLM(model="gpt-4", api_key="api.key")  # type: ignore
        assert isinstance(llm, LiteLLM)
        assert llm.model_name == "gpt-4"
```

File: src/distilabel/llm/litellm.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+    Attributes:
+        model: the model name to use for the LLM e.g. "gpt-3.5-turbo", "gpt-4", etc.
+        api_key: the API key to authenticate the requests to the OpenAI API.
+        verbose: whether to log the LiteLLM client's logs. Defaults to `False`.
+    """
+
+    model: str
+    verbose: bool = False
+    _aclient: Optional["callable"] = PrivateAttr(...)
```
Comment (alvarobartt) [inline src/distilabel/llm/litellm.py:38]: To be compliant with the rest of the `LLM` implementations (split between the `PrivateAttr` and the stardard ones

```suggestion
    model: str
    verbose: bool = False

    _aclient: Optional["callable"] = PrivateAttr(...)
```

**Pull request merged**